### PR TITLE
Datetime64 fixes

### PR DESCRIPTION
**Pull request by philippjfr**
Matplotlib does not currently support plotting numpy datetime64 types directly. This PR adds a compatibility layer ensuring these types are converted appropriately for plotting in matplotlib.

### CODE REVIEW

File: holoviews/plotting/mpl/chart.py [hunk 1]
```diff
@@ -61,6 +63,14 @@ def get_data(self, element, ranges, style):
         ys = element.dimension_values(1)
         return (xs, ys), style, {}
 
+    def init_artists(self, ax, plot_args, plot_kwargs):
+        xs, ys = plot_args
+        if xs.dtype.kind == 'M':
```
Comment (jlstevens) [inline holoviews/plotting/mpl/chart.py]: Might be worth adding a comment mentioning what the 'M' code means. I assume it is the numpy datetime64 type?

File: holoviews/plotting/util.py [hunk 2]
```diff
@@ -270,3 +272,10 @@ def dim_axis_label(dimensions, separator=', '):
     if not isinstance(dimensions, list): dimensions = [dimensions]
     return separator.join([safe_unicode(d.pprint_label)
                            for d in dimensions])
+
+def dt64_to_dt(dt64):
```
Comment (jlstevens) [inline holoviews/plotting/util.py]: Seems like a useful utility!

**Comment by jlstevens** [comment]
Looks good. Ready to merge?

**Comment by philippjfr** [comment]
Think there might be a few more places where datetime64 support can be improved. Will investigate a bit more.

**Comment by philippjfr** [comment]
I've now overhauled the PR again. The main changes here are that you can now register datetime formatters for datetime64 and datetime types, which means you can now do this:

```python
%%output size=100
%%opts Curve [xrotation=15 xticks=5]
import numpy as np
from datetime import date
import pandas as pd

start = date(2012, 1, 15)
end = date(2012, 1, 30)
hv.Dimension.type_formatters[np.datetime64] = '%m/%d'
hv.Curve((pd.date_range(start, end, freq='D'), np.random.rand(16)))
```

Additionally I fixed various bugs in xarray date range handling and made formatting of grid tick labels consistent with everything else. Ready to merge now, but there's a definite possibility tests will break due to Grid tick formatting changes.

**Comment by philippjfr** [comment]
Ready to review, updated test data will pass shortly.

File: holoviews/plotting/mpl/plot.py [hunk 3]
```diff
@@ -262,9 +262,6 @@ class GridPlot(CompositePlot):
     show_legend = param.Boolean(default=False, doc="""
         Legends add to much clutter in a grid and are disabled by default.""")
 
-    tick_format = param.String(default="%.2f", doc="""
```
Comment (jlstevens) [inline holoviews/plotting/mpl/plot.py]: So this parameter is now deprecated?

Comment (philippjfr) [inline holoviews/plotting/mpl/plot.py]: Hmm, yes, we have a system for tick formatting using ``Dimension.value_format`` and ``Dimension.type_formatters``, so this is redundant. You think we need to go through a deprecation cycle for this?

File: holoviews/core/dimension.py [hunk 4]
```diff
@@ -178,7 +179,11 @@ def pprint_value(self, value):
             if callable(formatter):
                 return formatter(value)
             elif isinstance(formatter, basestring):
-                if re.findall(r"\{(\w+)\}", formatter):
+                if isinstance(value, dt.datetime):
+                    return value.strftime(formatter)
+                elif isinstance(value, np.datetime64):
+                    return dt64_to_dt(value).strftime(formatter)
+                elif re.findall(r"\{(\w+)\}", formatter):
                     return formatter.format(value)
```
Comment (jlstevens) [inline holoviews/core/dimension.py]: So the datetime64 objects will use datetime style formatting? That is fine, assuming datetime64 doesn't have its own formatting conventions.

**Comment by jlstevens** [comment]
Looks good. I've made two comments and once you have addressed those, I am happy to merge.

File: holoviews/core/dimension.py (hunk 4 above)
Comment (philippjfr) [inline holoviews/core/dimension.py]: Yes, I think that's nice behavior, datetime64 can't be formatted directly so it converts to regular datetime first and then applies the formatter.

File: holoviews/plotting/mpl/plot.py (hunk 3 above)
Comment (jlstevens) [inline holoviews/plotting/mpl/plot.py]: It will at least need to be mentioned in the CHANGELOG. Other than that I think it is fine.

**Pull request merged**